### ISSUE TITLE
Global Styles: Allow saving, importing, exporting, copying from another theme without switching to it

### ISSUE THREAD

**Issue by jasmussen** [WordPress/gutenberg#45371]
In Global Styles > Browse styles, it should be possible to import, export, and save styles. The interface should accommodate all these flows:

1. Allow saving user variations for each bundled style (i.e. customized “dark” variation) (same as #38333)
2. Load style variations from a file
3. Provide access to styles from other themes without switching to them (related to #45452)
4. Provide an implicit warning about losing style customizations (related to #45210, #48407, #40561)
5. Allow exporting, deleting styles

The following mockups take a stab at the above, and may in some cases replace them. Here's importing a file:

<img width="3136" alt="Import styles from file" src="https://user-images.githubusercontent.com/1204802/199941872-25df2e11-8771-4de2-acb1-974426bd05a4.png">

* Import file exists in ellipsis menu
* Plus exists to create a style from scratch (separate future feature)

Styles from other themes can be surfaced:

<img width="1856" alt="Styles from other themes you have installed can also be leveraged" src="https://user-images.githubusercontent.com/1204802/199941943-9606da9b-da4b-4977-93fa-4ab4a60565a6.png">

* The "Theme styles" section acts as a ToolsPanel, showing by default the active theme styles, letting you toggle on styles from other installed themes.

Saving customizations:

<img width="1856" alt="Saving customizations" src="https://user-images.githubusercontent.com/1204802/199941970-9b4fb64a-d346-438b-b24b-29f1fb2e3e1d.png">

* Activating a style from a theme doesn't automatically save it as a custom style.
* But making any customizations to the style makes it show up under "Custom styles", with a blue dot. If you save using Multi Entity saving, or using the dropdown, the blue dot disappears.
* The ellipsis under custom styles allow saving, deleting, exporting.

As a starting point, styles imported or copied from other themes retain their stylenames from those sources. But in the future, we could explore a create from scratch + rename flow, such as this:

<img width="3728" alt="Creating from scratch" src="https://user-images.githubusercontent.com/1204802/199942007-70bfbf42-fb37-41e0-8400-46a72b4b2924.png">

---

Issue updated Nov 3. 

<details><summary>Initial proposal</summary>

This issue has a great deal of overlap with #38333, but takes a slightly different approach in that it is meant to include a flow for importing and exporting from files. The goal is to use the same interface to accomplish three separate flows:

1. Allow saving user variations for each bundled style (i.e. customized “dark” variation)
2. Load style variations from a file (import)
3. Copy global styles from another theme without switching to that theme (import)
5. Export, to enable items 2 and 3

Incidentally, these flows may also solve #45210, and depending on feedback, we may merge this issue into 38333, or vice versa.

In the first case, a user has picked a style variation that's bundled with a theme, but has made no other customizations to it:

<img width="1856" alt="User has no customizations" src="https://user-images.githubusercontent.com/1204802/198572876-9b58d465-07f7-46d1-83a9-fa437d3ce447.png">

* Chosen style shows as active in the list of bundled theme styles
* An ellipsis menu allows importing, exporting, and saving a style

If customizations are made to a style, or if a style is imported:

<img width="3040" alt="User has customizations" src="https://user-images.githubusercontent.com/1204802/198572962-d6d0e24b-6d73-4fa1-bd51-e0fc22990440.png">

* Customized style shows up in "Unsaved styles" section, clearly delineated as separate from the them bundled styles
* Ellipsis menu shows uption to save or delete
* If saved, the style will appear in a new section, called "Custom styles"
* If further customizations are made, the "Unsaved styles" section appears again

This is a first iteration, let me know your thoughts.

<details>

**Comment by jameskoster** [comment]
This looks like a good first iteration!

Should it be possible to export a single variation? I'm curious what that would look like. Similarly should it be possible to delete custom styles? I can see that section becoming a bit unwieldy otherwise.

Do we definitely need a separate 'Custom styles' section? I wonder if we could do something similar to templates, where customised theme files are marked in the UI, and it is possible to 'clear customisations'.

In a similar line of thinking, instead of an 'unsaved styles' section, could we just mark the currently selected style in a way to indicate its unsaved state?

Not something to implement here, but I mention it in case it inspires any relevant thoughts: in the future it might be nice to have a "New style" option which is essentially a blank slate. It may also be something you can trigger manually after customising an existing variation... essentially a "Save as..." option.

**Comment by jasmussen** [comment]
> Should it be possible to export a single variation? I'm curious what that would look like. Similarly should it be possible to delete custom styles? I can see that section becoming a bit unwieldy otherwise.

There's a delete button in the last mockup. Each action is meant to apply to the currently active/selected style. But perhaps the context isn't clear enough? Maybe we need ellipses on each style then? 🤔 

> Do we definitely need a separate 'Custom styles' section? I wonder if we could do something similar to templates, where customised theme files are marked in the UI, and it is possible to 'clear customisations'.

Might work, yes! Except ... 

> In a similar line of thinking, instead of an 'unsaved styles' section, could we just mark the currently selected style in a way to indicate its unsaved state?

There's a throughline here. By having separate Custom and Unsaved Styles sections, you'd never lose your work by switching to another, which is currently the case. 

I like your idea of taking inspiration from how it works in template management, though, seems like there could be a solution there.

**Comment by jameskoster** [comment]
> But perhaps the context isn't clear enough? 

Ah! That wasn't obvious to me. Since the ellipsis is next to the "Theme styles" heading, I expected the actions would be applied to all.

> There's a throughline here. By having separate Custom and Unsaved Styles sections, you'd never lose your work by switching to another, which is currently the case.

Of course, good point. But I think the main reason that's an issue currently is because the UI doesn't make it clear when your changes will be lost. There might be other ways around that, I'll mock something up quickly to test.

**Comment by jameskoster** [comment]
Here's what I was thinking:

<img width="1362" alt="Screenshot 2022-11-01 at 09 23 40" src="https://user-images.githubusercontent.com/846565/199202176-eb951df2-576f-4e38-ae66-1ac6b56a977f.png">

The main trade-off is that you cannot switch between a theme style, and a customised version of that same theme style, without reverting it. I'm not sure it's worth consolidating all the options into a single group after all...

**Comment by jasmussen** [comment]
The other tradeoff is the added noise of the style name, which is already integrated into the focused style. I wonder if the ellipsis can't be integrated as well? It would be additional focus stops, two per item, but it might be worth it? 

I still think there's some value in having unsaved styles show up in a brand new category, for a couple of reasons:

* In many cases you'll just pick a style and be happy.
* In other cases when you _do_ choose to customize a style, this fact is such a big deal that IMO it deserves separation

Mainly though, by making it a separate style, it can be "implicitly saved". Kind of like a draft in Figma. It wouldn't disappear just because you selected another style, you could always go back to it.

But perhaps "Unsaved styles" and "Custom styles" could be combined into a single category, rather than being two? Let me know what you think and I can take a stab.

**Comment by jameskoster** [comment]
> But perhaps "Unsaved styles" and "Custom styles" could be combined into a single category, rather than being two? Let me know what you think and I can take a stab.

Could be worth a try, with a detail to signify unsaved changes.

**Comment by jasmussen** [comment]
Alright, here's a new iteration based on your work, and back towards that of Channing in https://github.com/WordPress/gutenberg/issues/38333#issuecomment-1075727679, namely a horizontal ellipsis to balance utility with compression. User has no customizations:

<img width="1856" alt="User has no customizations" src="https://user-images.githubusercontent.com/1204802/199235219-f002d865-90e3-4a4a-bc00-d3c8074c5b5c.png">

User has customizations, and saves a style:

<img width="2448" alt="User has customizations" src="https://user-images.githubusercontent.com/1204802/199235181-8d347378-8634-42b7-8317-b03da08045ca.png">

**Comment by jameskoster** [comment]
Nice. The ellipsis looks a little awkward but I don't have a better idea right now. 

For the custom styles what is the blue dot highlighting?

Including 'Save style' is interesting. Contextual saving isn't a feature we have anywhere else but it might be nice to begin thinking about that. On a similar note, how do you feel about providing a 'Discard unsaved changes' action? I think that might be useful too.

**Comment by scruffian** [comment]
I had a slightly different but connected idea in https://github.com/WordPress/gutenberg/issues/45452

**Comment by jasmussen** [comment]
Excellent feedback, thank you all. Here's an iteration based on your thoughts, but which encompasses the idea of importing styles as outlined in 45452. Mockup:

<img width="3136" alt="Import styles from other themes" src="https://user-images.githubusercontent.com/1204802/199478770-e917f514-f683-49ea-9a2a-9cec1aace500.png">

* "Custom Styles" becomes a permanent fixture, even if you have none. This affords a plus button to import.
* Any theme with style variations that you have installed, will surface its style variations in the modal that opens. Pick and choose what to import.
* All imported styles will show up in "Custom styles".

You can also import from file, shown here importing "tt2-style-mint.json":

<img width="2544" alt="Import from file" src="https://user-images.githubusercontent.com/1204802/199478951-419b4862-0f38-45b4-ab1b-f9a2521813c8.png">

* The option to import from style sits in an ellipsis dropdown and opens an upload modal.
* Imported styles are appended after other styles in the "Custom styles" section.

Saving customizations:

<img width="1264" alt="Saving customizations" src="https://user-images.githubusercontent.com/1204802/199479135-73a1ec7f-8a61-4447-a157-e054e1e2781d.png">

* If you customize a theme style and save changes, a new style will appear in "Custom styles".
* If you customize it further, unsaved changes are denoted with a dot. 
* The ellipsis menu contains options to save or discard changes, or to duplicate or "download" (née "export") a style, as well as delete it.

What do you think?

**Comment by jameskoster** [comment]
This is really taking shape quickly!

The + button could easily be expanded to include an option to create a custom style from scratch, nice :) 

> All imported styles will show up in "Custom styles".

This was a _little_ unexpected to me. My instinct is that "custom styles" feel a bit different to something that comes from a theme you have installed. Also, what happens if you activate a theme whose styles you've already imported, do they then appear in both the "Theme styles" _and_ the "Custom styles" sections? 

Possibly a terrible idea, but instead of "Theme styles" you could have sections for each theme, IE "Twenty Twenty-Two Styles" and "Twenty Twenty-Three Styles" based on which ones you've imported. 

<img width="284" alt="Screenshot 2022-11-02 at 17 09 08" src="https://user-images.githubusercontent.com/846565/199555533-8014a9e9-bed8-4337-ae72-bd9a23b9d360.png">


Or even crazier... maybe you don't even need to "import". We could use the design tools UI to enable toggling of installed theme styles:

<img width="283" alt="Screenshot 2022-11-02 at 17 00 57" src="https://user-images.githubusercontent.com/846565/199554884-52523994-6395-4928-8cd3-24a03324bac5.png">

One final small point, I didn't see a way to rename custom styles, do we need that?

**Comment by jasmussen** [comment]
> This was a little unexpected to me. My instinct is that "custom styles" feel a bit different to something that comes from a theme you have installed. Also, what happens if you activate a theme whose styles you've already imported, do they then appear in both the "Theme styles" and the "Custom styles" sections?

I can see people initially wanting to just import everything to have choice, and then deciding more isn't necessarily better and needing to delete. Yes, when imported like this, they'd be duplicate.

I like the idea of just toggling the availability of styles like that, let's try it. I do think there are some technical questions that will likely emerge on what importing actually means, because we probably still need some way to literally copy over a style from a theme, so you can then go ahead and delete it. 

Let me update the mockups and issue, and loop in a dev for clarity! Thanks

**Comment by jasmussen** [comment]
Actually let me get one more round of feedback, because it remains somewhat complex. Here's a basic flow of importing from a file:

<img width="3136" alt="Import styles from file" src="https://user-images.githubusercontent.com/1204802/199699800-b1eabf44-595d-49eb-b29a-66779d633cb6.png">

* Import file exists in ellipsis menu.
* Plus remains for a future "create from scratch" feature.

You can surface styles from other installed themes:

<img width="1856" alt="Styles from other themes you have installed can also be leveraged" src="https://user-images.githubusercontent.com/1204802/199700002-bbc6e32e-77fc-4171-aef4-122c18bf2444.png">

* Theme styles acts as a ToolsPanel, showing by default the active theme styles, but you can toggle on styles from other installed themes.

Saving customizations:

<img width="1856" alt="Saving customizations" src="https://user-images.githubusercontent.com/1204802/199700161-27b0a806-bb88-4121-a613-e53bc0ecc0ae.png">

* Activating a style from a theme doesn't automatically save it as a custom style. 
* But making any customizations to the style makes it show up under "Custom styles", with a blue dot. If you save using Multi Entity saving, or using the dropdown, the blue dot disappears.
* The ellipsis under custom styles allow saving, deleting, exporting.

**A question**: Is it weird that theme styles has a panel style with a separator, but custom styles do not? Since there's already a lot of noise, I wanted to avoid borders if possible. Here's another option:

<img width="672" alt="alt" src="https://user-images.githubusercontent.com/1204802/199700716-8a2ad82a-982c-42c0-850e-8f8d0e9a7715.png">

**Comment by jameskoster** [comment]
This is looking good to me :) The only missing piece so far as I can see is the handling of custom style names. How do you see that working? Or do you think they just adopt the same name as their theme style counterpart? That may be fine to begin with, but worth noting that when we add the ability to create a style from scratch then naming/renaming will be required.

I don't have a strong feeling on the separator, but conceptually I do consider user-generated styles (regardless of whether they're based on theme styles initially) to be significantly different to bundled theme styles, so perhaps it is warranted?

**Comment by jasmussen** [comment]
> This is looking good to me :) The only missing piece so far as I can see is the handling of custom style names. How do you see that working? Or do you think they just adopt the same name as their theme style counterpart? That may be fine to begin with, but worth noting that when we add the ability to create a style from scratch then naming/renaming will be required.

Good call. My instinct would be that they adopt the names they are imported or copied with, and can't be edited until we also look at the "Create new" flow, which might be as simple as this:
<img width="3728" alt="Creating from scratch" src="https://user-images.githubusercontent.com/1204802/199709610-7e92aaa9-b6ab-4931-a500-df3f9b7d88ee.png">

**Comment by jasmussen** [comment]
This issue has been updated and labelled "Needs Dev", as we've settled on a design that could likely work for a first iteration.

**Comment by carolinan** [comment]
Currently, WP_REST_Global_Styles_Controller is limited to the active theme so it needs to be updated.

I assume that once a style from an installed but not active theme is selected and you save it, the intention is that it gets saved in the active themes global style option in the database.

I also assume that we need both the base theme.json and the style variation from the installed but not active themes.

How do we handle font files that are used in style variations, and are placed in the installed but not active theme?
- It is possible that a style variation can be saved and the theme that has the font files is deleted.
- It is also possible that the read/write permissions are not enough to allow copying the font files from that theme and place them in a wp-content/fonts folder or similar.

**Comment by jameskoster** [comment]
Yes, I suspect both of those are possible albeit rare occurrences. Perhaps we need to update the font selection control so that it displays an error/warning if the selected font cannot be found?

**Comment by carolinan** [comment]
Also should this include child themes?
I guess there are 3 options,
ignore child themes,
use a possibly partial style variation,
and fetching all 3: parent theme.json, child theme theme.json, child theme variations.

**Comment by ribaricplusplus** [comment]
I am planning to take care of Save and Delete for user variations in #46952, covering point 1. from the main issue comment. I will probably not work on the other points but I'm assigning this to myself temporarily because working on import / export in parallel would be unfeasible.

**Comment by annezazu** [comment]
Thanks so much for jumping into this! Please share as you go where you might need help or reviews.

**Comment by ribaricplusplus** [comment]
@annezazu It's 1100+ lines of code... Definitely going to need some reviews haha. Should be done in a day or two. I see that you're very active, I'll ping you when it's done, maybe you can help me find some people to review it.

**Comment by unprintedch** [comment]
Great work, that is exactly what I was hoping for here #60386.
Is there some PR we can test?

**Comment by annezazu** [comment]
@scruffian great job getting https://github.com/WordPress/gutenberg/pull/63318 merged to create a new public function to expose variations from other themes. Where does that leave us in terms of what's next? I see some performance issues mentioned in the PR reviews btw.

**Comment by carolinan** [comment]
@youknowriad @jasmussen @ellatrix

Hi
@afercia @SergeyBiryukov and I were wondering if there is any progress on points 2: Load style variations from a file and 5: Allow exporting, deleting styles.

We are wondering how this can be aligned with the proposal in https://github.com/WordPress/gutenberg/issues/66950.
-If I have edited global styles, and I switch themes, shouldn't I be able to re-use the changes I made to the global styles, 
without a complicated import/export flow?

**Comment by youknowriad** [comment]
> If I have edited global styles, and I switch themes, shouldn't I be able to re-use the changes I made to the global styles,
without a complicated import/export flow?

You should, the issue right now is that the global styles only save the "user modifications" and they get merged on top of the theme provided theme.json file.

Indeed, I'm hoping that we can use the work in https://github.com/WordPress/gutenberg/issues/66950 to inform what comes next for both style variations and patterns.

**Comment by carolinan** [comment]
I am not sure that the combined theme.json and global styles changes would always be the best choice.

I see many users asking to re-use the CSS from the Additional CSS field.
